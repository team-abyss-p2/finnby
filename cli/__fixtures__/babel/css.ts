// @ts-expect-error unresolved self import
import { css } from "@team-abyss-p2/finnby";

const bgBlue = css`
    background-color: blue;
`;
