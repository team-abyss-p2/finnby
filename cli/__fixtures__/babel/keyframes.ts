/* eslint-disable */
// @ts-nocheck

import { keyframes } from "@team-abyss-p2/finnby";

const fadeIn = keyframes`
    from {
        opacity: 0;
    }
    to {
        opacity: 1;
    }
`;
