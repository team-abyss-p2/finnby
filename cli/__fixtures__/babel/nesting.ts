/* eslint-disable */
// @ts-nocheck

import { styled } from "@team-abyss-p2/finnby";

const Base = styled.Panel`
    background-color: blue;
`;

const Component = styled(Base)`
    color: red;
`;
