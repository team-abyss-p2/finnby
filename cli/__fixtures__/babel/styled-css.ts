// @ts-expect-error unresolved self import
import { css, styled } from "@team-abyss-p2/finnby";

const bgBlue = css`
    background-color: blue;
`;

const Component = styled.Panel`
    @extend ${bgBlue};
    color: red;
`;
