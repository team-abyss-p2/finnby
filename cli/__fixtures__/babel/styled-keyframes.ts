// @ts-expect-error unresolved self import
import { keyframes, styled } from "@team-abyss-p2/finnby";

const fadeIn = keyframes`
    from {
        opacity: 0;
    }
    to {
        opacity: 1;
    }
`;

const Component = styled.Panel`
    animation: ${fadeIn} 500ms;
`;
