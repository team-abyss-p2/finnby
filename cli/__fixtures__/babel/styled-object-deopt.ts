// @ts-expect-error unresolved self import
import { styled } from "@team-abyss-p2/finnby";

const Component = styled.Panel(
    // @ts-expect-error missing proptype
    (props) => ({
        backgroundColor: props.bg,
    }),
);
