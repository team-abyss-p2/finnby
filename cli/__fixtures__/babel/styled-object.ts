// @ts-expect-error unresolved self import
import { styled } from "@team-abyss-p2/finnby";

const Component = styled.Panel({
    backgroundColor: "red",
});
