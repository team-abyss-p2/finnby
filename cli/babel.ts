import { PluginObj, PluginPass, NodePath, types as t } from "@babel/core";

interface StyledState extends PluginPass {
    cssImport?: string;
    styledImport?: string;
}

export function styledPlugin(
    uid: string,
    styles: string[],
): PluginObj<StyledState> {
    return {
        visitor: {
            ImportDeclaration(path, state) {
                const source = path.get("source");
                if (source.node.value !== "@team-abyss-p2/finnby") {
                    return;
                }

                for (const specifier of path.get("specifiers")) {
                    if (!specifier.isImportSpecifier()) {
                        continue;
                    }

                    const imported = specifier.get("imported");
                    if (!imported.isIdentifier()) {
                        continue;
                    }

                    const local = specifier.get("local");
                    switch (imported.node.name) {
                        case "css":
                            state.cssImport = local.node.name;
                            continue;
                        case "styled":
                            state.styledImport = local.node.name;
                            continue;
                    }
                }
            },
            TaggedTemplateExpression(path, state) {
                const tag = path.get("tag");
                const quasi = path.get("quasi");

                const newNode = tryFoldConstantStyle(
                    uid,
                    styles,
                    state,
                    tag,
                    quasi,
                );

                if (!newNode) {
                    return;
                }

                path.replaceWith(newNode);
            },
            CallExpression(path, state) {
                const callee = path.get("callee");
                if (!callee.isExpression()) {
                    return;
                }

                const [style, ...args] = path.get("arguments");
                if (!style.isExpression()) {
                    return;
                }

                const newNode = tryFoldConstantStyle(
                    uid,
                    styles,
                    state,
                    callee,
                    style,
                    args.map((arg) => arg.node),
                );

                if (!newNode) {
                    return;
                }

                path.replaceWith(newNode);
            },
        },
    };
}

type Argument =
    | t.Expression
    | t.SpreadElement
    | t.JSXNamespacedName
    | t.ArgumentPlaceholder;

function tryFoldConstantStyle(
    uid: string,
    styles: string[],
    state: StyledState,
    expr: NodePath<t.Expression>,
    style: NodePath<t.Expression>,
    additionalArgs: Argument[] = [],
) {
    let styledProp: NodePath<t.Expression | t.PrivateName> | undefined;
    let styledArgs: NodePath<Argument>[] | undefined;

    if (expr.isMemberExpression()) {
        // styled.Panel() / styled.Panel``
        const object = expr.get("object");
        if (!object.isIdentifier()) {
            return;
        }
        if (object.node.name !== state.styledImport) {
            return;
        }

        styledProp = expr.get("property");
    } else if (expr.isCallExpression()) {
        // styled(Panel)() / styled(Panel)``
        const callee = expr.get("callee");
        if (!callee.isIdentifier()) {
            return;
        }
        if (callee.node.name !== state.styledImport) {
            return;
        }

        styledArgs = expr.get("arguments");
    } else if (expr.isIdentifier()) {
        // css() / css``
        if (expr.node.name !== state.cssImport) {
            return;
        }
    } else {
        return;
    }

    const code = style.evaluate();
    if (!code.confident) {
        return;
    }

    let newCallee: t.Expression;
    if (styledProp && state.styledImport) {
        newCallee = t.memberExpression(
            t.memberExpression(
                t.identifier(state.styledImport),
                t.identifier("static"),
            ),
            styledProp.node,
        );
    } else if (styledArgs && state.styledImport) {
        newCallee = t.callExpression(
            t.memberExpression(
                t.identifier(state.styledImport),
                t.identifier("static"),
            ),
            styledArgs.map((arg) => arg.node),
        );
    } else if (state.cssImport) {
        newCallee = t.memberExpression(
            t.identifier(state.cssImport),
            t.identifier("static"),
        );
    } else {
        return;
    }

    const className = `_${uid}_${styles.length}`;

    if (typeof code.value === "string") {
        styles.push(`.${className}{${code.value}}`);
    } else {
        throw new Error("Not implemented");
    }

    return t.callExpression(newCallee, [
        t.stringLiteral(className),
        ...additionalArgs,
    ]);
}
