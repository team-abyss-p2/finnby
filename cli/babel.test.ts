import pluginTester from "babel-plugin-tester";
import { styledPlugin } from "./babel";

const styles: string[] = [];

beforeEach(() => {
    styles.length = 0;
});

pluginTester({
    pluginName: "styled plugin",
    plugin: styledPlugin({
        uid: "_uid",
        styles,
    }),

    filename: __filename,
    snapshot: true,

    tests: {
        "should inline static keyframes": {
            fixture: "__fixtures__/babel/keyframes.ts",
            teardown() {
                expect(styles).toMatchSnapshot();
            },
        },
        "should inline static css": {
            fixture: "__fixtures__/babel/css.ts",
            teardown() {
                expect(styles).toMatchSnapshot();
            },
        },

        "should generate a class for static styled component templates": {
            fixture: "__fixtures__/babel/styled-template.ts",
            teardown() {
                expect(styles).toMatchSnapshot();
            },
        },
        "should generate a class for static styled component objects": {
            fixture: "__fixtures__/babel/styled-object.ts",
            teardown() {
                expect(styles).toMatchSnapshot();
            },
        },

        "should inline keyframe names in styled components": {
            fixture: "__fixtures__/babel/styled-keyframes.ts",
            teardown() {
                expect(styles).toMatchSnapshot();
            },
        },
        "should inline css class names in styled components": {
            fixture: "__fixtures__/babel/styled-css.ts",
            teardown() {
                expect(styles).toMatchSnapshot();
            },
        },

        "should deopt on non-static styled component template strings": {
            fixture: "__fixtures__/babel/styled-template-deopt.ts",
            teardown() {
                expect(styles).toMatchSnapshot();
            },
        },
        "should deopt on non-static styled component objects": {
            fixture: "__fixtures__/babel/styled-object-deopt.ts",
            teardown() {
                expect(styles).toMatchSnapshot();
            },
        },

        "should handle styled components being used in selectors": {
            fixture: "__fixtures__/babel/styled-selector.ts",
            teardown() {
                expect(styles).toMatchSnapshot();
            },
        },
    },
});
