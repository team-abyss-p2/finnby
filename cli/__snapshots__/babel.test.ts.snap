// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`styled plugin should deopt on non-static styled component objects 1`] = `Array []`;

exports[`styled plugin should deopt on non-static styled component objects: should deopt on non-static styled component objects 1`] = `

// @ts-expect-error unresolved self import
import { styled } from "@team-abyss-p2/finnby";

const Component = styled.Panel(
    // @ts-expect-error missing proptype
    (props) => ({
        backgroundColor: props.bg,
    }),
);

      ↓ ↓ ↓ ↓ ↓ ↓

// @ts-expect-error unresolved self import
import { styled } from "@team-abyss-p2/finnby";
const Component = styled.Panel(
    // @ts-expect-error missing proptype
    (props) => ({
        backgroundColor: props.bg,
    }),
);


`;

exports[`styled plugin should deopt on non-static styled component template strings 1`] = `Array []`;

exports[`styled plugin should deopt on non-static styled component template strings: should deopt on non-static styled component template strings 1`] = `

// @ts-expect-error unresolved self import
import { styled } from "@team-abyss-p2/finnby";

const Component = styled.Panel\`
    color: \${
        // @ts-expect-error missing proptype
        (props) => props.color
    };
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

// @ts-expect-error unresolved self import
import { styled } from "@team-abyss-p2/finnby";
const Component = styled.Panel\`
    color: \${
        // @ts-expect-error missing proptype
        (props) => props.color
    };
\`;


`;

exports[`styled plugin should generate a class for static styled component objects 1`] = `
Array [
  ._uid_0_Component{background-color: red;},
]
`;

exports[`styled plugin should generate a class for static styled component objects: should generate a class for static styled component objects 1`] = `

// @ts-expect-error unresolved self import
import { styled } from "@team-abyss-p2/finnby";

const Component = styled.Panel({
    backgroundColor: "red",
});

      ↓ ↓ ↓ ↓ ↓ ↓

// @ts-expect-error unresolved self import
import { styled } from "@team-abyss-p2/finnby";
const Component = /*#__PURE__*/ styled.static.Panel("_uid_0_Component");


`;

exports[`styled plugin should generate a class for static styled component templates 1`] = `
Array [
  ._uid_0_Component{
    color: red;
},
]
`;

exports[`styled plugin should generate a class for static styled component templates: should generate a class for static styled component templates 1`] = `

// @ts-expect-error unresolved self import
import { styled } from "@team-abyss-p2/finnby";

const Component = styled.Panel\`
    color: red;
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

// @ts-expect-error unresolved self import
import { styled } from "@team-abyss-p2/finnby";
const Component = /*#__PURE__*/ styled.static.Panel("_uid_0_Component");


`;

exports[`styled plugin should handle styled components being used in selectors 1`] = `
Array [
  ._uid_0_ComponentA{
    color: red;
},
  ._uid_1_ComponentB{
    color: red;

    ._uid_0_ComponentA {
        color: blue;
    }
},
]
`;

exports[`styled plugin should handle styled components being used in selectors: should handle styled components being used in selectors 1`] = `

// @ts-expect-error unresolved self import
import { styled } from "@team-abyss-p2/finnby";

const ComponentA = styled.Panel\`
    color: red;
\`;

const ComponentB = styled.Panel\`
    color: red;

    \${ComponentA} {
        color: blue;
    }
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

// @ts-expect-error unresolved self import
import { styled } from "@team-abyss-p2/finnby";
const ComponentA = /*#__PURE__*/ styled.static.Panel("_uid_0_ComponentA");
const ComponentB = /*#__PURE__*/ styled.static.Panel("_uid_1_ComponentB");


`;

exports[`styled plugin should inline css class names in styled components 1`] = `
Array [
  ._uid_0_bgBlue{
    background-color: blue;
},
  ._uid_1_Component{
    @extend ._uid_0_bgBlue;
    color: red;
},
]
`;

exports[`styled plugin should inline css class names in styled components: should inline css class names in styled components 1`] = `

// @ts-expect-error unresolved self import
import { css, styled } from "@team-abyss-p2/finnby";

const bgBlue = css\`
    background-color: blue;
\`;

const Component = styled.Panel\`
    @extend \${bgBlue};
    color: red;
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

// @ts-expect-error unresolved self import
import { styled } from "@team-abyss-p2/finnby";
const bgBlue = /*#__PURE__*/ "_uid_0_bgBlue";
const Component = /*#__PURE__*/ styled.static.Panel("_uid_1_Component");


`;

exports[`styled plugin should inline keyframe names in styled components 1`] = `
Array [
  @keyframes _uid_0_fadeIn{
    from {
        opacity: 0;
    }
    to {
        opacity: 1;
    }
},
  ._uid_1_Component{
    animation: _uid_0_fadeIn 500ms;
},
]
`;

exports[`styled plugin should inline keyframe names in styled components: should inline keyframe names in styled components 1`] = `

// @ts-expect-error unresolved self import
import { keyframes, styled } from "@team-abyss-p2/finnby";

const fadeIn = keyframes\`
    from {
        opacity: 0;
    }
    to {
        opacity: 1;
    }
\`;

const Component = styled.Panel\`
    animation: \${fadeIn} 500ms;
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

// @ts-expect-error unresolved self import
import { styled } from "@team-abyss-p2/finnby";
const fadeIn = /*#__PURE__*/ "_uid_0_fadeIn";
const Component = /*#__PURE__*/ styled.static.Panel("_uid_1_Component");


`;

exports[`styled plugin should inline static css 1`] = `
Array [
  ._uid_0_bgBlue{
    background-color: blue;
},
]
`;

exports[`styled plugin should inline static css: should inline static css 1`] = `

// @ts-expect-error unresolved self import
import { css } from "@team-abyss-p2/finnby";

const bgBlue = css\`
    background-color: blue;
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

// @ts-expect-error unresolved self import
import "@team-abyss-p2/finnby";
const bgBlue = /*#__PURE__*/ "_uid_0_bgBlue";


`;

exports[`styled plugin should inline static keyframes 1`] = `
Array [
  @keyframes _uid_0_fadeIn{
    from {
        opacity: 0;
    }
    to {
        opacity: 1;
    }
},
]
`;

exports[`styled plugin should inline static keyframes: should inline static keyframes 1`] = `

// @ts-expect-error unresolved self import
import { keyframes } from "@team-abyss-p2/finnby";

const fadeIn = keyframes\`
    from {
        opacity: 0;
    }
    to {
        opacity: 1;
    }
\`;

      ↓ ↓ ↓ ↓ ↓ ↓

// @ts-expect-error unresolved self import
import "@team-abyss-p2/finnby";
const fadeIn = /*#__PURE__*/ "_uid_0_fadeIn";


`;
